examples/pytorch/diffpool/model/dgl_layers/gnn.py

```diff
@@ -101,27 +102,13 @@ def __init__(self, input_dim, assign_dim, output_feat_dim,
         self.reg_loss.append(EntropyLoss())
 
     def forward(self, g, h):
-        feat = self.feat_gc(g, h)
-        assign_tensor = self.pool_gc(g, h)
+        feat = self.feat_gc(g, h)  # size = (sum_N, F_out), sum_N is num of nodes in this batch
         device = feat.device
-        assign_tensor_masks = []
-        batch_size = len(g.batch_num_nodes())
-        for g_n_nodes in g.batch_num_nodes():
-            mask = torch.ones((g_n_nodes,
-                               int(assign_tensor.size()[1] / batch_size)))
-            assign_tensor_masks.append(mask)
-        """
-        The first pooling layer is computed on batched graph.
-        We first take the adjacency matrix of the batched graph, which is block-wise diagonal.
-        We then compute the assignment matrix for the whole batch graph, which will also be block diagonal
-        """
-        mask = torch.FloatTensor(
-            block_diag(
-                *
-                assign_tensor_masks)).to(
-            device=device)
-        assign_tensor = masked_softmax(assign_tensor, mask,
-                                       memory_efficient=False)
+        assign_tensor = self.pool_gc(g, h)  # size = (sum_N, N_a), N_a is num of nodes in pooled graph.
+        assign_tensor = F.softmax(assign_tensor, dim=1)
+        assign_tensor = torch.split(assign_tensor, g.batch_num_nodes().tolist())
+        assign_tensor = torch.block_diag(*assign_tensor)  # size = (sum_N, batch_size * N_a)
```

Comment:
Seems `torch.block_diag` returns a dense tensor.

Another strategy is to split `assign_tensor` and `feat`, loop the `matmul`, and finally concatenate back.

Is this strategy much slower than the call in L112? If the speed is comparable I'd say we avoid the creation of a large block tensor.

Reply:
It is worth trying this method on node features. However, this method is expensive for pooling edges because we can only get one big matrix from graphs in dgl (no matter dense or sparse).

Reply:
OK that makes sense, so that means the computation on L114 - L115 are needing this diagonal.